Add tag support for announcements

nworbiii opened the pull request:
## Hey, I just made a Pull Request!

Provide finer-grained filtering discovery of Announcements based upon Tags.

Announcements currently define a Category to which they are attributed. However, a category may be too broad. Tags would provide the means to further filter announcements.

Closes [#2977](https://github.com/backstage/community-plugins/issues/2977)

![new-announcement](https://github.com/user-attachments/assets/88d83be0-1644-454b-8538-f4e5e4cafe10)
![home](https://github.com/user-attachments/assets/de58af59-be15-496f-bde3-3f4a7ca614f7)
![announcements-list](https://github.com/user-attachments/assets/38a5984f-b4e2-4a5c-839a-0c22480f17e9)
![tags-list](https://github.com/user-attachments/assets/510a0916-66eb-44b0-bcec-f00c13cc1239)

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

kurtaking commented:
I'm dealing with an illness while taking care of a sick little one. It'll be a few days before I can review.

kurtaking commented:
@gaelgoth - do you have time (and want) to pull this and test it?

@nworbiii - I've reviewed half the code itself. Hoping to have it done early this week as I get back in the swing of things.

gaelgoth commented:
Thank you, @nworbiii, for the PR!

I ran a few tests on my end. I noticed that when filling the tags field and pressing <kbd>Enter</kbd>, the tag isn't registered in the database, which results in a 500 error when submitting the announcement.

**This issue isn't introduced by this feature**. I've observed the same behavior with the Category field as well. However, it's less noticeable there since it's a single-entry field.

I find that for the tags field, the feedback that you need to click on “Create ...” to save an entry in the list is less obvious

![announcements_tags](https://github.com/user-attachments/assets/c191aba9-95ac-4957-8daf-3c986d7bf75b)

I'm open to other alternatives, perhaps it would be better to save the array once the user has entered all the tags?

gaelgoth reviewed (commented):
I detected this issue with a PostgreSQL database. 

When the providing an empty tag array: 

```json
{"category":"test","start_at":"2025-04-16","tags":[],"title":"Test","excerpt":"No tags","body":"Hello world!","active":true,"publisher":"user:development/guest"}
```

I got the following error from the backend: 

```log
stack="error: insert into \"announcements\" (\"active\", \"body\", \"category\", \"created_at\", \"excerpt\", \"id\", \"publisher\", \"start_at\", \"tags\", \"title\") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) - malformed array literal: \"[]\"\n    at Parser.parseErrorMessage (/home/ggt/cloned-repo/temp-cloned/community-plugins/workspaces/announcements/node_modules/pg-protocol/src/parser.ts:368:69)\n
```
I didn't encounter this error with SQLite

nworbiii commented on workspaces/announcements/plugins/announcements-backend/src/service/persistence/AnnouncementsDatabase.ts:
```diff
@@ -150,9 +180,15 @@ export class AnnouncementsDatabase {
+      if (tags?.length) {
+        // For SQLite JSON, use LIKE for each tag
+        tags.forEach(tag => {
+          qb.whereRaw(`tags LIKE ?`, [`%${tag}%`]);
```
I switched to using whereLike and changed tags in announcementUpsertToDB to use JSON.stringify([]) instead of undefined when announcement.tags doesn't exist

nworbiii commented:
> Thank you, @nworbiii, for the PR!
> 
> I ran a few tests on my end. I noticed that when filling the tags field and pressing Enter, the tag isn't registered in the database, which results in a 500 error when submitting the announcement.
> 
> **This issue isn't introduced by this feature**. I've observed the same behavior with the Category field as well. However, it's less noticeable there since it's a single-entry field.
> 
> I find that for the tags field, the feedback that you need to click on “Create ...” to save an entry in the list is less obvious
> 
> ![announcements_tags](https://private-user-images.githubusercontent.com/55390114/434241195-c191aba9-95ac-4957-8daf-3c986d7bf75b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MTgzNjQsIm5iZiI6MTc0NTQxODA2NCwicGF0aCI6Ii81NTM5MDExNC80MzQyNDExOTUtYzE5MWFiYTktOTVhYy00OTU3LThkYWYtM2M5ODZkN2JmNzViLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDE0MjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OWI0MTk0ZTc2MDVmMmM5ZWVhMjA2YmQ4ODZhNGRlMGJkZGUyOWU3ODc2YTZjYzNiZTVlMzIxNWI4NjRiNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GaSCsJJmy4mi7xrlbbbEd-oUNaeYDlaI24PvUFHZ0Ys) [ ![announcements_tags](https://private-user-images.githubusercontent.com/55390114/434241195-c191aba9-95ac-4957-8daf-3c986d7bf75b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MTgzNjQsIm5iZiI6MTc0NTQxODA2NCwicGF0aCI6Ii81NTM5MDExNC80MzQyNDExOTUtYzE5MWFiYTktOTVhYy00OTU3LThkYWYtM2M5ODZkN2JmNzViLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDE0MjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OWI0MTk0ZTc2MDVmMmM5ZWVhMjA2YmQ4ODZhNGRlMGJkZGUyOWU3ODc2YTZjYzNiZTVlMzIxNWI4NjRiNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GaSCsJJmy4mi7xrlbbbEd-oUNaeYDlaI24PvUFHZ0Ys) ](https://private-user-images.githubusercontent.com/55390114/434241195-c191aba9-95ac-4957-8daf-3c986d7bf75b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MTgzNjQsIm5iZiI6MTc0NTQxODA2NCwicGF0aCI6Ii81NTM5MDExNC80MzQyNDExOTUtYzE5MWFiYTktOTVhYy00OTU3LThkYWYtM2M5ODZkN2JmNzViLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDE0MjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OWI0MTk0ZTc2MDVmMmM5ZWVhMjA2YmQ4ODZhNGRlMGJkZGUyOWU3ODc2YTZjYzNiZTVlMzIxNWI4NjRiNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GaSCsJJmy4mi7xrlbbbEd-oUNaeYDlaI24PvUFHZ0Ys) [ ](https://private-user-images.githubusercontent.com/55390114/434241195-c191aba9-95ac-4957-8daf-3c986d7bf75b.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MTgzNjQsIm5iZiI6MTc0NTQxODA2NCwicGF0aCI6Ii81NTM5MDExNC80MzQyNDExOTUtYzE5MWFiYTktOTVhYy00OTU3LThkYWYtM2M5ODZkN2JmNzViLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDE0MjEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3OWI0MTk0ZTc2MDVmMmM5ZWVhMjA2YmQ4ODZhNGRlMGJkZGUyOWU3ODc2YTZjYzNiZTVlMzIxNWI4NjRiNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GaSCsJJmy4mi7xrlbbbEd-oUNaeYDlaI24PvUFHZ0Ys)
> 
> I'm open to other alternatives, perhaps it would be better to save the array once the user has entered all the tags?

The tags now do not save until the AnnouncementForm is submitted

gaelgoth commented on workspaces/announcements/plugins/announcements-backend/src/service/persistence/AnnouncementsDatabase.ts:
```diff
@@ -95,6 +117,9 @@ const announcementUpsertToDB = (
     created_at: announcement.created_at.toSQL()!,
     active: announcement.active,
     start_at: announcement.start_at.toSQL()!,
+    tags: announcement.tags
+      ? JSON.stringify(announcement.tags)
+      : JSON.stringify([]),
```
I did some tests with PostgresQL. The database expects an array, but the value being inserted is a malformed array literal `[\"test\"]`. PostgreSQL expects array literals to be formatted as `{value1,value2}` instead of JSON-like strings
It works with both psql and sqlite:
```suggestion
    tags: announcement.tags ? `{${announcement.tags.join(',')}}` : '{}'
```

gaelgoth reviewed (changes requested):
Your branch is currently a bit behind the base branch, could you please update it when you get a chance?

Also, with the recent [Backstage upgrade](https://github.com/backstage/community-plugins/pull/3668) to v1.38, you’ll need to [adjust your React component](https://backstage.io/docs/tutorials/jsx-transform-migration) to align with the new JSX transform.

Once that’s done, I’ll do a full review again. Thanks! 👍🏾

nworbiii commented:
> Your branch is currently a bit behind the base branch, could you please update it when you get a chance?
> 
> Also, with the recent [Backstage upgrade](https://github.com/backstage/community-plugins/pull/3668) to v1.38, you’ll need to [adjust your React component](https://backstage.io/docs/tutorials/jsx-transform-migration) to align with the new JSX transform.
> 
> Once that’s done, I’ll do a full review again. Thanks! 👍🏾

Should be good now.

vinzscam commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -326,6 +339,77 @@ export async function createRouter(
+      }
+
+      const tag = {
+        ...req.body,
```
we are missing some validation here and in the endpoint below

nworbiii replied:
Added additional checks

nworbiii commented:
- added validation to endpoints
- changed table names from announcements_tags to tags
- used chips array for TagsInput
- added additional handling of duplicates in TagsInput
- fixed TagsContent styling to match CategoriesContent
- fixed regression with tags not displaying in announcements

nworbiii commented on workspaces/announcements/plugins/announcements/src/components/AnnouncementForm/TagsInput.tsx:
```diff
@@ -0,0 +1,170 @@
+  const { t } = useAnnouncementsTranslation();
+  const createLabel = t('announcementForm.tagsInput.create');
+
+  const existingTags = useMemo(() => {
```
Please test again

gaelgoth replied:
I was unable to delete a tag, even though it wasn't being used in any announcement.

If a tag is not associated with any announcement, it should be deletable. This should follow the same logic as category deletion.

Not that I did my test with PostgreSQL DB. Here are the steps to reproduce:

- Go to http://localhost:3000/announcements/admin/create
- Create a new tag
- Attempt to delete the created tag
- Backend throws an error when attempting to delete the unused tag:

```log
2025-05-09T11:26:30.536Z announcements error Filtered internal error with logId=2171d5ec670831aa983f from response select "id", "publisher", "announcements"."title", "excerpt", "body", "category", "created_at", "categories"."title" as "category_title", "active", "start_at", "tags" from "announcements" left join "categories" on "announcements"."category" = "categories"."slug" where "tags" like $1 order by "created_at" desc - operator does not exist: text[] ~~ unknown type="errorHandler" logId="2171d5ec670831aa983f" length=203 name="error" severity="ERROR" code="42883" detail=undefined hint="No operator matches the given name and argument types. You might need to add explicit type casts." position="277" internalPosition=undefined internalQuery=undefined where=undefined schema=undefined table=undefined column=undefined dataType=undefined constraint=undefined file="parse_oper.c" line="647" routine="op_error" stack="error: select \"id\", \"publisher\", \"announcements\".\"title\", \"excerpt\", \"body\", \"category\", \"created_at\", \"categories\".\"title\" as \"category_title\", \"active\", \"start_at\", \"tags\" from \"announcements\" left join \"categories\" on \"announcements\".\"category\" = \"categories\".\"slug\" where \"tags\" like $1 order by \"created_at\" desc - operator does not exist: text[] ~~ unknown\n    at Parser.parseErrorMessage (/home/github/community-plugins/workspaces/announcements/node_modules/pg-protocol/src/parser.ts:368:69)\n    at Parser.handlePacket (/home/github/community-plugins/workspaces/announcements/node_modules/pg-protocol/src/parser.ts:187:21)\n    at Parser.parse (/home/github/community-plugins/workspaces/announcements/node_modules/pg-protocol/src/parser.ts:102:30)\n    at Socket.<anonymous> (/home/github/community-plugins/workspaces/announcements/node_modules/pg-protocol/src/index.ts:7:48)\n    at Socket.emit (node:events:518:28)\n    at addChunk (node:internal/streams/readable:561:12)\n    at readableAddChunkPushByteMode (node:internal/streams/readable:512:3)\n    at Readable.push (node:internal/streams/readable:392:5)\n    at TCP.onStreamRead (node:internal/stream_base_commons:189:23)"
```

nworbiii replied:
Hello, how can I get authenticated with the postgres backend up? Or disable auth locally. My app-config.yaml has auth.providers.guest = {} but the authorize call fails

gaelgoth replied:
Hey @nworbiii,

You can leave your `app-config.yaml` as is.

Just create a new `app-config.local.yaml` with the following content:

```yaml
backend:
  baseUrl: http://localhost:7007
  listen:
    port: 7007
  database:
    # Duplicate this block in app-config.local.yaml for Postgres
    client: pg
    connection:
      host: ${POSTGRES_HOST}
      port: ${POSTGRES_PORT}
      user: ${POSTGRES_USER}
      password: ${POSTGRES_PASSWORD}
```

For database setup, follow the instructions here:  https://github.com/backstage/community-plugins/blob/main/workspaces/announcements/plugins/announcements-backend/README.md?plain=1#L42-L64

Make sure your environment variables are set with `.env` file values.
Once that’s done, you're all set,  you can start the plugin using postgres

nworbiii replied:
I tested deletion of a tag not tied to an announcement and of a tag tied to an announcement with the postgres backend. Please test agian. Thanks!

gaelgoth replied:
This issue is now fixed. However, I've encountered a regression on the create form.

 Steps to reproduce the issue (still with Postgres):

1. Create a new tag from the "Tags" tab, I entered "new tag"
2. Create a new announcement, fill in the form, and select the newly created tag
3. Submit the form.

After submitting, I receive a 409 error with the following response:

```json
{
  "error": "Tag already exists"
}

gaelgoth reviewed (commented):
I did a few tests on my side and didn’t notice any bugs.  

Please rebase your branch on `main` and resolve the conflicts, we added a new field in the form you may adapt it for tags. 
 
I left few comments. After that, we’ll be ready to ship 🚢

nworbiii commented on workspaces/announcements/plugins/announcements/src/components/AnnouncementsCard/AnnouncementsCard.tsx:
```diff
@@ -151,6 +152,21 @@ export const AnnouncementsCard = ({
+                            component={Link}
+                            to={`${announcementsLink()}?tags=${tag.slug}`}
+                            clickable
+                            style={{ marginRight: 4, marginBottom: 4 }}
```
Hi @gaelgoth, I addressed all of your comments (minus the batching of calls) and rebased my branch on main.

nworbiii commented on workspaces/announcements/plugins/announcements/src/components/Admin/AnnouncementsContent/AnnouncementsContent.tsx:
```diff
@@ -256,10 +264,11 @@ export const AnnouncementsContent = () => {
       render: rowData => {
         return (
-          <>
+          <div style={{ display: 'flex', flexDirection: 'row' }}>
```
Fixed this one. To clarify, you mean batching the calls on announcements? I can make that change if it's a dealbreaker.

gaelgoth replied:
Nice! I will review and test this change on my side

nworbiii commented on workspaces/announcements/plugins/announcements-backend/src/service/persistence/AnnouncementsDatabase.ts:
```diff
@@ -179,18 +217,36 @@ export class AnnouncementsDatabase {
+
+    const announcementRows = await queryBuilder;
+
+    const results = await Promise.all(
```
This should be resolved now.

gaelgoth commented on workspaces/announcements/plugins/announcements-backend/db/migrations/20250327141958_add_announcements_tags.js:
```diff
@@ -0,0 +1,33 @@
+  });
+
+  await knex.schema.alterTable('announcements', table => {
+    table.specificType('tags', 'TEXT[]');
```
During my tests, I encountered errors because the raw SQL ANY operator was used to fetch tags, which isn’t supported by SQLite. 

I’m wondering if you should refactoring a bit and set the column to use .json, which is supported by both databases

```suggestion
table.json('tags').comment('Array of tag slugs');
```

This will also avoid to have different logic between the DB like this: https://github.com/backstage/community-plugins/pull/3530/files#diff-5d8b4b2837d2ea813d9d4462c4b4bda3fdf797bac98a20f0bafa3abdb479f7b7R55-R64

kurtaking commented:
How are we feeling about merging once the API report is fixed? @gaelgoth

nworbiii commented:
> Hi @nworbiii, there’s still a bit of refactoring needed to fully support table.json('tags'). For example, we could remove this part into a single implementation by removing `if (isPostgres)`: [#3530 (files)](https://github.com/backstage/community-plugins/pull/3530/files#diff-5d8b4b2837d2ea813d9d4462c4b4bda3fdf797bac98a20f0bafa3abdb479f7b7R55-R64).
> 
> I let you double check that all CRUD are OK with the json format.
> 
> To fix formatting, please run `yarn prettier:fix`.
> 
> I’ll focus on your PR so we can ship this feature soon! 🚀

Ah, I'd forgotten to pull the suggested change to my local that was causing the prettier check to fail.

I did the refactor. It seems to be working in both sqlite and pg but please test again.

gaelgoth reviewed (commented):
Thank you for the json column refactoring. 

I tested it on my end with both SQLite and PostgreSQL, and everything works without any issues.

I’ve left a few comments with minor changes to help align the code with the existing style. 

Please resolve the conflicts on this PR so we can proceed with merging your work

nworbiii commented on workspaces/announcements/plugins/announcements-backend/src/service/persistence/TagsDatabase.ts:
```diff
@@ -0,0 +1,70 @@
+
+  async tagBySlug(slug: string): Promise<Tag | undefined> {
+    return this.db('tags').where({ slug }).first();
+  }
```
@gaelgoth Ready for review again

nworbiii replied:
@gaelgoth Ready for review again

gaelgoth replied:
Hi @nworbiii. 

The following test is failing:

```log
FAIL @backstage-community/plugin-announcements-backend plugins/announcements-backend/src/service/persistence/TagsDatabase.test.ts
  ● TagsDatabase › should handle attempting to insert a duplicate tag gracefully

    expect(received).rejects.toThrow()

    Received promise resolved instead of rejected
    Resolved to value: undefined

      131 |     });
      132 |
    > 133 |     await expect(
          |           ^
      134 |       tagsStore.insert({
      135 |         slug: 'unique-tag',
      136 |         title: 'Different Title',

      at expect (../../../node_modules/expect/build/index.js:113:15)
      at Object.expect (service/persistence/TagsDatabase.test.ts:133:11)
````

nworbiii replied:
Thanks. Forgot to run those. Should be fixed.

gaelgoth reviewed (approved):
All good for me! 

Hello @kurtaking, can you please take a quick look 👀 ?

kurtaking commented:
> Ah, it looks like we also need your review on this, @vinzscam

Reviewing now. If things loook good to you and I, I think he will be okay with us dismissing. His changes were from early on and I'm pretty sure they are addressed.

kurtaking commented:
I am failing to create a new announcement with added tags locally.  Can someone else test locally with a clean instance using postgres?

gaelgoth commented:
Hi @nworbiii, when you have a chance, could you please take a look at @kurtaking feedback?

nworbiii commented:
> I am failing to create a new announcement with added tags locally. Can someone else test locally with a clean instance using postgres?
> 
> **edit**: should we be calling `JSON.stringify` on the tags before inserting into the DB? And it might be worth storing all tags as an array, even when there is only one.

Hi @kurtaking. I connected to psql and dropped all of the backstage related databases and then ran db:setup from inside plugin-announcements-backend, saw that everything was wiped, and was able to create a new announcement with a tag. Could you try that? Or if someone else can try to reproduce.

I can make the those extra changes if they are blockers to getting the PR merged.

gaelgoth commented:
Hi @nworbiii,

I tested it again on my end, no issues found. Let’s go ahead and ship this feature 🚢
